062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+They will be able to change the topic replication factor by simply updating the `KafkaTopic` resource, which may also happen through a GitOps pipeline.
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can reuse some of the existing logic and classes.
+
+At runtime, the Topic Operator watches and periodically checks all managed and non-paused `KafkaTopic` resources.
+Among the other configurations, the reconciliation logic also search for `.spec.replicas` update by comparing with the actual topic replication factor.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- Pending: `.spec.replicas` is updated, but the configuration change request is not sent to Cruise Control (`generation` > `observedGeneration`).
+- Ongoing: `.spec.replicas` is updated and Cruise Control has accepted the request, but the task is not completed (`generation` == `observedGeneration`).
+- Failed: `.spec.replicas` is updated with an invalid value, or there is a hard goal violation, or Cruise Control is not available (`generation` == `observedGeneration`).
+
+To request a replication factor change, the Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.
+For that reason, we introduce the following environment variables, which will be set automatically by the Cluster Operator, or manually by the user in case of standalone deployment.
+
+| Name                                   | Type    | Default | Description                                           |
+|:---------------------------------------|---------|---------|-------------------------------------------------------|
+| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled         |
+| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running             |
+| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                               |
+| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                   |
+| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether Cruise Control SSL encryption is enabled      |
+| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Cruise Control Basic authorization is enabled |
```

Comment:
Do you really want to make this configurable? It might be hard to test.

Reply:
When I added this, I was thinking about the use case of a standalone deployment user with a CC instance running without basic authentication. Do we want to support this?

Reply:
Well, It is not that it is wrong. But you have to think about the test strategy, dcoumentation, helping the users etc. Ss the regular STs do not test it and docs do not cover much about it either. So if nothing else, it is wroth mentioning these things and covering them clearly in the proposal.